Wine (with mingw support)

**Project description**
There is a great wine package currently in NixOS but it lacks mingw support for the DLLs.

This is needed for a few games including World of Warcraft it seems. Basically we need to build WINE with "--with-mingw" compile flag. The problem I have here is - I have no idea how to actually include the mingw compiler as a dependency to WINE.

Any help here would be appreciated.

**Metadata**

* homepage URL: https://www.winehq.org/ | http://www.mingw.org/
* source URL: n/a
* license: n/a
* platforms: linux

@7c6f434c I'm not too sure what closure means in this context D: What I will say, however, is that this is OFF by default in most distros so I think it should be added as an option and not enabled by default - just as you suggested.

@MetaDark Your changes got me a lot further in to the build but I believe I am missing one critical change - the build runs correctly and starts building DLLs via mingw but then fails out with:

`/build/wine-5.20/dlls/msvcp60/tests/ios.c:1470:9: error: format not a string literal and no format arguments [-Werror=format-security]`

Actual overlay pkg:
```
self: super:

{
  wine-mingw = super.wineWowPackages.full.overrideDerivation (oldAttrs: rec {
    name = "wine-${version}";
    version = "5.20";

    buildScript = ./builder-wow.sh;

    wineRelease = "staging";

    src = super.fetchurl {
      url = "https://dl.winehq.org/wine/source/5.x/wine-${version}.tar.xz";
      sha256 = "0746j29six9cjgxp9pxfw1r67xhi39kvmqaq5yhvmgdyrvw24mcg";
    };

    # Fixes "Compiler cannot create executables" building with wineWow
    strictDeps = true;

    nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [
      # needed for wine32 or wineWow
      super.pkgsCross.mingw32.windows.crossThreadsStdenv.cc

      # needed for wine64 or wineWow
      super.pkgsCross.mingwW64.windows.crossThreadsStdenv.cc
    ];

    # Fixes -Werror=format-security
    hardeningDisable = oldAttrs.hardeningDisable ++ [ "format" ];
  });
}
```

> @MetaDark Your changes got me a lot further in to the build but I believe I am missing one critical change - the build runs correctly and starts building DLLs via mingw but then fails out with:
> 
> `/build/wine-5.20/dlls/msvcp60/tests/ios.c:1470:9: error: format not a string literal and no format arguments [-Werror=format-security]`
> 
> Actual overlay pkg (including the fix for exactly this error from your code linked above.) I wonder if mingw is ignoring the hardeningDisable flag or is using defaults somewhere? We're so close to fixing this I can feel it! :+1:

@cawilliamson Hmm, it worked for me when specifying `hardeningDisable`. Maybe there is a problem with using `overrideDerivation`? The Nixpkgs manual strongly recommends using `overrideAttrs` instead of `overrideDerivation`. I think it's very likely that's the cause: https://nixos.org/manual/nixpkgs/stable/#sec-pkg-overrideDerivation

@7c6f434c I'll gather information about the closure sizes for each configuration (wine32, wine64, wineWow) and I'll also look into creating a PR that adds a new support flag.

--- a/pkgs/top-level/wine-packages.nix
+++ b/pkgs/top-level/wine-packages.nix
@@ -27,6 +27,7 @@ rec {
     xineramaSupport = true;
     xmlSupport = true;
     sdlSupport = true;
+    mingwSupport = true;
   };
 
   full = base.override {

